File: ch15/Chapter15-CustomProviders-CS/App_Code/XmlMembershipProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Hosting;
using System.Web.Security;
using System.Xml;
using System.Xml.Linq;

namespace Chapter15_CustomProviders_CS.App_Code
{
    public class XmlMembershipProvider : MembershipProvider
    {
        private string _AppName;
        private Dictionary<string, MembershipUser> _MyUsers;
        private string _FileName;

        public override string ApplicationName
        {
            get
            {
                return _AppName;
            }
            set
            {
                _AppName = value;
            }
        }

        public override void Initialize(string name,
            System.Collections.Specialized.NameValueCollection config)
        {
            base.Initialize(name, config);

            _AppName = config["applicationName"];

            if (String.IsNullOrEmpty(_AppName))
            {
                _AppName = "/";
            }

            _FileName = config["xmlUserDatabaseFile"];

            if (String.IsNullOrEmpty(_FileName))
            {
                _FileName = "/App_Data/Users.xml";
            }
        }

        public override bool ChangePassword(string username, string oldPassword, string newPassword)
        {
            return false;
        }

        public override bool ChangePasswordQuestionAndAnswer(string username, string password, string newPasswordQuestion, string newPasswordAnswer)
        {
            throw new NotImplementedException();
        }

        public override MembershipUser CreateUser(string username, string password, string email, string passwordQuestion, string passwordAnswer, bool isApproved, object providerUserKey, out MembershipCreateStatus status)
        {
            throw new NotImplementedException();
        }

        public override bool DeleteUser(string username, bool deleteAllRelatedData)
        {
            throw new NotImplementedException();
        }

        public override bool EnablePasswordReset
        {
            get { throw new NotImplementedException(); }
        }

        public override bool EnablePasswordRetrieval
        {
            get { throw new NotImplementedException(); }
        }

        public override MembershipUserCollection FindUsersByEmail(string emailToMatch, int pageIndex, int pageSize, out int totalRecords)
        {
            throw new NotImplementedException();
        }

        public override MembershipUserCollection FindUsersByName(string usernameToMatch, int pageIndex, int pageSize, out int totalRecords)
        {
            throw new NotImplementedException();
        }

        public override MembershipUserCollection GetAllUsers(int pageIndex, int pageSize, out int totalRecords)
        {
            throw new NotImplementedException();
        }

        public override int GetNumberOfUsersOnline()
        {
            throw new NotImplementedException();
        }

        public override string GetPassword(string username, string answer)
        {
            throw new NotImplementedException();
        }

        public override MembershipUser GetUser(string username, bool userIsOnline)
        {
            throw new NotImplementedException();
        }

        public override MembershipUser GetUser(object providerUserKey, bool userIsOnline)
        {
            throw new NotImplementedException();
        }

        public override string GetUserNameByEmail(string email)
        {
            throw new NotImplementedException();
        }

        public override int MaxInvalidPasswordAttempts
        {
            get { throw new NotImplementedException(); }
        }

        public override int MinRequiredNonAlphanumericCharacters
        {
            get { throw new NotImplementedException(); }
        }

        public override int MinRequiredPasswordLength
        {
            get { throw new NotImplementedException(); }
        }

        public override int PasswordAttemptWindow
        {
            get { throw new NotImplementedException(); }
        }

        public override MembershipPasswordFormat PasswordFormat
        {
            get { throw new NotImplementedException(); }
        }

        public override string PasswordStrengthRegularExpression
        {
            get { throw new NotImplementedException(); }
        }

        public override bool RequiresQuestionAndAnswer
        {
            get { throw new NotImplementedException(); }
        }

        public override bool RequiresUniqueEmail
        {
            get { throw new NotImplementedException(); }
        }

        public override string ResetPassword(string username, string answer)
        {
            throw new NotImplementedException();
        }

        public override bool UnlockUser(string userName)
        {
            throw new NotImplementedException();
        }

        public override void UpdateUser(MembershipUser user)
        {
            throw new NotImplementedException();
        }

        public override bool ValidateUser(string username, string password)
        {
            if (String.IsNullOrEmpty(username) || String.IsNullOrEmpty(password))
            {
                return false;
            }

            try
            {
                ReadUserFile();

                MembershipUser mu;

                if (_MyUsers.TryGetValue(username.ToLower(), out mu))
                {
                    if (mu.Comment == password)
                    {
                        return true;
                    }
                }

                return false;
            }
            catch (Exception ex)
            {
                throw new Exception(ex.Message.ToString());
            }
        }

        private void ReadUserFile()
        {
            if (_MyUsers == null)
            {
                lock (this)
                {
                    _MyUsers = new Dictionary<string, MembershipUser>();
                    var query = from users in 
                                XElement.Load(HostingEnvironment.MapPath(_FileName)).Elements("User")
                                select users;

                    foreach (var user in query)
                    {
                        MembershipUser mu = new MembershipUser(Name,
                            user.Element("Username").Value,
                            null,
                            user.Element("Email").Value,
                            String.Empty,
                            user.Element("Password").Value,
                            true,
                            false,
                            DateTime.Parse(user.Element("DateCreated").Value),
                            DateTime.Now,
                            DateTime.Now,
                            DateTime.Now,
                            DateTime.Now);

                        _MyUsers.Add(mu.UserName.ToLower(), mu);
                    }
                }
            }
        }
    }
}

File: ch15/Chapter15-CustomProviders-VB/App_Code/XmlMembershipProvider.vb
Imports System.Xml
Imports System.Web.Hosting

Public Class XmlMembershipProvider
    Inherits MembershipProvider

    Private _AppName As String
    Private _MyUsers As Dictionary(Of String, MembershipUser)
    Private _FileName As String

    Public Overrides Property ApplicationName() As String
        Get
            Return _AppName
        End Get
        Set(ByVal value As String)
            _AppName = value
        End Set
    End Property

    Public Overrides Sub Initialize(ByVal name As String, _
        ByVal config As System.Collections.Specialized.NameValueCollection)
        MyBase.Initialize(name, config)

        _AppName = config("applicationName")

        If (String.IsNullOrEmpty(_AppName)) Then
            _AppName = "/"
        End If

        _FileName = config("xmlUserDatabaseFile")

        If (String.IsNullOrEmpty(_FileName)) Then
            _FileName = "/App_Data/Users.xml"
        End If
    End Sub

    Public Overrides Function ChangePassword(username As String, oldPassword As String, newPassword As String) As Boolean
        Return False
    End Function

    Public Overrides Function ChangePasswordQuestionAndAnswer(username As String, password As String, newPasswordQuestion As String, newPasswordAnswer As String) As Boolean

    End Function

    Public Overrides Function CreateUser(username As String, password As String, email As String, passwordQuestion As String, passwordAnswer As String, isApproved As Boolean, providerUserKey As Object, ByRef status As MembershipCreateStatus) As MembershipUser

    End Function

    Public Overrides Function DeleteUser(username As String, deleteAllRelatedData As Boolean) As Boolean

    End Function

    Public Overrides ReadOnly Property EnablePasswordReset As Boolean
        Get

        End Get
    End Property

    Public Overrides ReadOnly Property EnablePasswordRetrieval As Boolean
        Get

        End Get
    End Property

    Public Overrides Function FindUsersByEmail(emailToMatch As String, pageIndex As Integer, pageSize As Integer, ByRef totalRecords As Integer) As MembershipUserCollection

    End Function

    Public Overrides Function FindUsersByName(usernameToMatch As String, pageIndex As Integer, pageSize As Integer, ByRef totalRecords As Integer) As MembershipUserCollection

    End Function

    Public Overrides Function GetAllUsers(pageIndex As Integer, pageSize As Integer, ByRef totalRecords As Integer) As MembershipUserCollection

    End Function

    Public Overrides Function GetNumberOfUsersOnline() As Integer

    End Function

    Public Overrides Function GetPassword(username As String, answer As String) As String

    End Function

    Public Overloads Overrides Function GetUser(providerUserKey As Object, userIsOnline As Boolean) As MembershipUser

    End Function

    Public Overloads Overrides Function GetUser(username As String, userIsOnline As Boolean) As MembershipUser

    End Function

    Public Overrides Function GetUserNameByEmail(email As String) As String

    End Function

    Public Overrides ReadOnly Property MaxInvalidPasswordAttempts As Integer
        Get

        End Get
    End Property

    Public Overrides ReadOnly Property MinRequiredNonAlphanumericCharacters As Integer
        Get

        End Get
    End Property

    Public Overrides ReadOnly Property MinRequiredPasswordLength As Integer
        Get

        End Get
    End Property

    Public Overrides ReadOnly Property PasswordAttemptWindow As Integer
        Get

        End Get
    End Property

    Public Overrides ReadOnly Property PasswordFormat As MembershipPasswordFormat
        Get

        End Get
    End Property

    Public Overrides ReadOnly Property PasswordStrengthRegularExpression As String
        Get

        End Get
    End Property

    Public Overrides ReadOnly Property RequiresQuestionAndAnswer As Boolean
        Get

        End Get
    End Property

    Public Overrides ReadOnly Property RequiresUniqueEmail As Boolean
        Get

        End Get
    End Property

    Public Overrides Function ResetPassword(username As String, answer As String) As String

    End Function

    Public Overrides Function UnlockUser(userName As String) As Boolean

    End Function

    Public Overrides Sub UpdateUser(user As MembershipUser)

    End Sub

    Public Overrides Function ValidateUser(username As String, password As String) _
        As Boolean
        If (String.IsNullOrEmpty(username) Or String.IsNullOrEmpty(password)) Then
            Return False
        End If

        Try
            ReadUserFile()

            Dim mu As MembershipUser

            If (_MyUsers.TryGetValue(username.ToLower(), mu)) Then
                If (mu.Comment = password) Then
                    Return True
                End If
            End If

            Return False
        Catch ex As Exception
            Throw New Exception(ex.Message.ToString())
        End Try
    End Function

    Private Sub ReadUserFile()
        If (_MyUsers Is Nothing) Then
            SyncLock (Me)
                _MyUsers = New Dictionary(Of String, MembershipUser)()
                Dim query = From users In _
                            XElement.Load( _
                                HostingEnvironment.MapPath(_FileName)).Elements("User") _
                            Select users

                For Each user In query
                    Dim mu As MembershipUser = New MembershipUser(Name, _
                       user.Element("Username").Value, _
                       Nothing, _
                       user.Element("Email").Value, _
                       String.Empty, _
                       user.Element("Password").Value, _
                       True, _
                       False, _
                       DateTime.Parse(user.Element("DateCreated").Value), _
                       DateTime.Now, _
                       DateTime.Now, _
                       DateTime.Now, _
                       DateTime.Now)

                    _MyUsers.Add(mu.UserName.ToLower(), mu)
                Next
            End SyncLock
        End If
    End Sub
End Class